Move `pal::env` to `std::sys::env_consts`

thaliaarchi opened the pull request:
Combine the `std::env::consts` platform implementations as a single file. Use the Unix file as the base, since it has 28 entries, and fold the 8 singleton platforms into it. The Unix file was roughly grouped into Linux, Apple, BSD, and everything else, roughly in alphabetical order. Alphabetically order them to make it easier to maintain and discard the Unix-specific groups to generalize it to all platforms.

I'd prefer to have no fallback implementation, as I consider it a bug; however TEEOS, Trusty, and Xous have no definitions here. Since they otherwise have `pal` abstractions, that indicates that there are several platforms without `pal` abstractions which are also missing here. To support unsupported, create a little macro to handle the fallback case and not introduce ordering between the `cfg`s like `cfg_if!`.

I've named the module `std::sys::env_consts`, because they are used in `std::env::consts` and I intend to use the name `std::sys::env` for the combination of `Args` and `Vars`.

cc @joboet @ChrisDenton

Tracked in #117276.

thaliaarchi committed:
Sort Unix env constants alphabetically by target_os

They were roughly grouped into Linux, Apple, BSD, and everything else,
roughly in alphabetical order. Alphabetically order them to make it
easier to maintain and discard the Unix-specific groups to generalize it
to all platforms.

thaliaarchi committed:
Combine env consts into std::sys::env_consts

thaliaarchi committed:
Handle unsupported fallback

thaliaarchi commented:
[WALI](https://doc.rust-lang.org/nightly/rustc/platform-support/wasm32-wali-linux.html) presents a complication, because it's both Linux and Wasm, but it's probably just as simple as `#[cfg(all(target_family = "wasm", not(target_os = "linux")))]`. The only others which don't gate on `target_os` are `#[cfg(all(target_vendor = "fortanix", target_env = "sgx"))]` and `#[windows]`. SGX is easy to verify, since there's only one target. For Windows, I've expanded `#[windows]` to `#[cfg(target_family = "windows")]`, since that's what it's a shorthand for, though `#[cfg(target_os = "windows")]` would be more clear. According to @ChrisDenton, there's probably not any cases where they differ.

joboet commented:
Sound and looks good to me!
@bors r+

bors commented:
:pushpin: Commit cac458c933a0e61c64a0e3296b995a2c39aa4303 has been approved by `joboet`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=joboet cac458c933a0e61c64a0e3296b995a2c39aa4303 -->
<!-- homu: {"type":"Approved","sha":"cac458c933a0e61c64a0e3296b995a2c39aa4303","approver":"joboet","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

matthiaskrgr commented:
@bors r-
https://github.com/rust-lang/rust/pull/140012#issuecomment-2816128811

```
2025-04-18T20:16:55.6950825Z
2025-04-18T20:16:55.8638464Z error[E0428]: the name `os` is defined multiple times
2025-04-18T20:16:55.8639221Z    --> library/std/src/sys/env_consts.rs:362:1
2025-04-18T20:16:55.8639750Z     |
2025-04-18T20:16:55.8640277Z 65  | pub mod os {
2025-04-18T20:16:55.8641094Z     | ---------- previous definition of the module `os` here
2025-04-18T20:16:55.8641700Z ...
2025-04-18T20:16:55.8642184Z 362 | pub mod os {
2025-04-18T20:16:55.8642884Z     | ^^^^^^^^^^ `os` redefined here
2025-04-18T20:16:55.8643648Z     |
2025-04-18T20:16:55.8644365Z     = note: `os` must be defined only once in the type namespace of this module
2025-04-18T20:16:55.8644825Z
2025-04-18T20:16:58.4808620Z For more information about this error, try `rustc --explain E0428`.
2025-04-18T20:16:58.5006460Z [RUSTC-TIMING] std test:false 2.826
2025-04-18T20:16:58.5009351Z warning: `std` (lib) generated 1 warning
2025-04-18T20:16:58.5010054Z error: could not compile `std` (lib) due to 1 previous error; 1 warning emitted
2025-04-18T20:16:58.5068975Z Build completed unsuccessfully in 0:00:34
2025-04-18T20:16:58.5157722Z   local time: Fri Apr 18 20:16:58 UTC 2025
2025-04-18T20:16:58.5715563Z   network time: Fri, 18 Apr 2025 20:16:58 GMT
``

thaliaarchi committed:
Use struct update syntax for some TargetOptions

thaliaarchi commented:
I've fixed Emscripten (which failed CI).

Also, I confirmed that `target_os = "emscripten"` always implies `target_arch = "wasm32" and that `target_family = "windows" iff `target_os = "windows`, so I switched those two to just `target_os` to make them more clear. Now, only Wasm and SGX aren't gated using `target_os` and I've similarly verified that they're mutually exclusive with the others.

```sh
$ rustc +nightly -Z unstable-options --print all-target-specs-json > targets.json
$ jq 'to_entries[] | select(.value | .os == "emscripten" and .arch != "wasm32")' targets.json
$ jq 'to_entries[] | select(.value | (."target-family" // [] | contains(["windows"])) != (.os == "windows"))' targets.json
$ jq 'map(select(."target-family" // [] | contains(["wasm"])).os) | unique' targets.json
[
  null,
  "emscripten",
  "linux",
  "unknown",
  "wasi"
]
$ jq '.[] | select(.vendor == "fortanix" or .env == "sgx").os' targets.json
"unknown"
```

In the process of this exploration, I found two tiny code simplifications in `rustc_targets` to match the style of other targets.

thaliaarchi commented:
The macro is now much simpler, while doing the same thing.

joboet commented:
Neat!
@bors r+

bors commented:
:pushpin: Commit 93fa96cfba126d542ebd33624e5f774d397dbf3f has been approved by `joboet`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=joboet 93fa96cfba126d542ebd33624e5f774d397dbf3f -->
<!-- homu: {"type":"Approved","sha":"93fa96cfba126d542ebd33624e5f774d397dbf3f","approver":"joboet","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
